relayer/chains/wasm/provider.go

```diff
@@ -569,42 +572,111 @@ func (p *Provider) fetchBlockMessages(ctx context.Context, heightInfo *types.Hei
 	)
 
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
 	for _, event := range p.eventList {
 		wg.Add(1)
-		go func(wg *sync.WaitGroup, searchParam types.TxSearchParam, messagesChan chan *coreTypes.ResultTxSearch, errorChan chan error) {
+		go func(wg *sync.WaitGroup, event sdkTypes.Event, searchParam types.TxSearchParam, messagesChan chan *coreTypes.ResultTxSearch, errorChan chan error) {
 			defer wg.Done()
-			searchParam.Events = append(searchParam.Events, event)
-			res, err := p.client.TxSearch(ctx, searchParam)
+
+			var (
+				localSearchParam = searchParam
+				localMessages    = new(coreTypes.ResultTxSearch)
+				zapFields        = []zap.Field{
+					zap.Uint64("start_height", localSearchParam.StartHeight),
+					zap.Uint64("end_height", localSearchParam.EndHeight),
+					zap.String("event", event.Type),
+				}
+			)
+			localSearchParam.Events = append(localSearchParam.Events, event)
+
+			err := retry.Retry(ctx, p.logger, func() error {
+				p.logger.Info("fetching block messages", zapFields...)
+				msgs, err := p.client.TxSearch(ctx, localSearchParam)
+				if err == nil {
+					p.logger.Info("fetched block messages", zapFields...)
+					localMessages = msgs
+				}
+				return err
+			}, zapFields)
 			if err != nil {
-				errorChan <- err
+				select {
+				case errorChan <- err:
+				default:
+				}
+				cancel()
 				return
 			}
-			if res.TotalCount > perPage {
-				for i := 2; i <= int(res.TotalCount/perPage)+1; i++ {
-					searchParam.Page = &i
-					resNext, err := p.client.TxSearch(ctx, searchParam)
+
+			if localMessages.TotalCount > perPage {
+				totalPages := (localMessages.TotalCount + perPage - 1) / perPage
+				for i := 2; i <= totalPages; i++ {
+					select {
+					case <-ctx.Done():
+						return
+					default:
+					}
+					p.logger.Info("fetching block messages", append(zapFields, zap.Int("page", i))...)
+					localSearchParam.Page = &i
+					err := retry.Retry(ctx, p.logger, func() error {
+						resNext, err := p.client.TxSearch(ctx, localSearchParam)
+						if err == nil {
+							localMessages.Txs = append(localMessages.Txs, resNext.Txs...)
+						}
+						return err
+					}, append(zapFields, zap.Int("page", i)))
 					if err != nil {
-						errorChan <- err
+						select {
+						case errorChan <- err:
+						default:
+						}
+						cancel()
 						return
 					}
-					res.Txs = append(res.Txs, resNext.Txs...)
 				}
 			}
-			messagesChan <- res
-		}(&wg, searchParam, messagesChan, errorChan)
+			messagesChan <- localMessages
+		}(&wg, event, searchParam, messagesChan, errorChan)
+	}
+
+	go func() {
+		wg.Wait()
+		close(messagesChan)
+		close(errorChan)
+	}()
+
+	var errors []error
+	for {
 		select {
-		case msgs := <-messagesChan:
-			messages.Txs = append(messages.Txs, msgs.Txs...)
-			messages.TotalCount += msgs.TotalCount
-		case err := <-errorChan:
-			p.logger.Error("failed to fetch block messages", zap.Error(err))
+		case msgs, ok := <-messagesChan:
```

Comment:
Can we simplify this using for loop over channel and checking the length of error channel to return early and make it more readable ??